File: android/src/main/java/com/margelo/nitro/nitrobiometrics/NitroBiometrics.kt
package com.margelo.nitro.nitrobiometrics

import android.app.KeyguardManager
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.os.Handler
import android.os.Looper
import androidx.annotation.Keep
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise

@DoNotStrip
@Keep
class NitroBiometrics : HybridNitroBiometricsSpec() {

  private val context: Context?
    get() = NitroModules.applicationContext ?: NitroBiometricsPackage.reactApplicationContext

  override fun canAuthenticate(allowDeviceCredentials: Boolean?): Promise<BiometricsStatus> {
    val promise = Promise<BiometricsStatus>()
    val ctx = context
    if (ctx == null) {
      promise.resolve(
        BiometricsStatus(
          isAvailable = false,
          biometryType = BiometryType.NONE,
          biometryTypes = emptyArray(),
          enrolled = false,
          isDeviceSecure = false,
          error = "contextUnavailable"
        )
      )
      return promise
    }

    val authenticators = if (allowDeviceCredentials == true) {
      BiometricManager.Authenticators.BIOMETRIC_STRONG or
        BiometricManager.Authenticators.BIOMETRIC_WEAK or
        BiometricManager.Authenticators.DEVICE_CREDENTIAL
    } else {
      BiometricManager.Authenticators.BIOMETRIC_STRONG or
        BiometricManager.Authenticators.BIOMETRIC_WEAK
    }

    val biometricManager = BiometricManager.from(ctx)
    val canAuthResult = biometricManager.canAuthenticate(authenticators)

    val pm = ctx.packageManager
    val biometryList = mutableListOf<BiometryType>()

    if (pm.hasSystemFeature(PackageManager.FEATURE_FINGERPRINT)) {
      biometryList.add(BiometryType.FINGERPRINT)
    }
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q && pm.hasSystemFeature(PackageManager.FEATURE_FACE)) {
      biometryList.add(BiometryType.FACE)
    }
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q && pm.hasSystemFeature(PackageManager.FEATURE_IRIS)) {
      biometryList.add(BiometryType.IRIS)
    }

    val primaryType = when {
      biometryList.contains(BiometryType.FACE) -> BiometryType.FACE
      biometryList.contains(BiometryType.FINGERPRINT) -> BiometryType.FINGERPRINT
      biometryList.contains(BiometryType.IRIS) -> BiometryType.IRIS
      else -> BiometryType.NONE
    }

    val keyguardManager = ctx.getSystemService(Context.KEYGUARD_SERVICE) as? KeyguardManager
    val isDeviceSecure = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
      keyguardManager?.isDeviceSecure ?: false
    } else {
      @Suppress("DEPRECATION")
      keyguardManager?.isKeyguardSecure ?: false
    }

    val isAvailable = canAuthResult == BiometricManager.BIOMETRIC_SUCCESS
    val enrolled = canAuthResult != BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED &&
      canAuthResult != BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE

    val error = when (canAuthResult) {
      BiometricManager.BIOMETRIC_SUCCESS -> null
      BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED -> "notEnrolled"
      BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE -> "notSupported"
      BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE -> "hardwareUnavailable"
      BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED -> "securityUpdateRequired"
      else -> "unknown"
    }

    val status = BiometricsStatus(
      isAvailable = isAvailable,
      biometryType = primaryType,
      biometryTypes = biometryList.toTypedArray(),
      enrolled = enrolled,
      isDeviceSecure = isDeviceSecure,
      error = error
    )

    promise.resolve(status)
    return promise
  }

  override fun authenticate(options: AuthenticateOptions): Promise<AuthenticateResult> {
    val promise = Promise<AuthenticateResult>()
    val reactContext = NitroModules.applicationContext ?: NitroBiometricsPackage.reactApplicationContext
    val currentActivity = reactContext?.currentActivity

    val fragmentActivity = currentActivity as? FragmentActivity
    if (fragmentActivity == null) {
      promise.resolve(
        AuthenticateResult(
          success = false,
          biometryType = null,
          error = "activityUnavailable",
          warning = "Current activity is not a FragmentActivity or is unavailable"
        )
      )
      return promise
    }

    val pm = fragmentActivity.packageManager
    val primaryType = when {
      Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q && pm.hasSystemFeature(PackageManager.FEATURE_FACE) -> BiometryType.FACE
      pm.hasSystemFeature(PackageManager.FEATURE_FINGERPRINT) -> BiometryType.FINGERPRINT
      Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q && pm.hasSystemFeature(PackageManager.FEATURE_IRIS) -> BiometryType.IRIS
      else -> null
    }

    Handler(Looper.getMainLooper()).post {
      try {
        val authenticators = if (options.allowDeviceCredentials == true) {
          BiometricManager.Authenticators.BIOMETRIC_STRONG or
            BiometricManager.Authenticators.BIOMETRIC_WEAK or
            BiometricManager.Authenticators.DEVICE_CREDENTIAL
        } else {
          BiometricManager.Authenticators.BIOMETRIC_STRONG or
            BiometricManager.Authenticators.BIOMETRIC_WEAK
        }

        val promptInfoBuilder = BiometricPrompt.PromptInfo.Builder()
          .setTitle(options.promptMessage)
          .setAllowedAuthenticators(authenticators)

        options.subtitle?.let { promptInfoBuilder.setSubtitle(it) }
        options.description?.let { promptInfoBuilder.setDescription(it) }

        if (options.confirmationRequired != null) {
          promptInfoBuilder.setConfirmationRequired(options.confirmationRequired)
        }

        if (options.allowDeviceCredentials != true) {
          promptInfoBuilder.setNegativeButtonText(options.cancelButtonText ?: "Cancel")
        }

        val promptInfo = promptInfoBuilder.build()
        val executor = ContextCompat.getMainExecutor(fragmentActivity)

        val biometricPrompt = BiometricPrompt(
          fragmentActivity,
          executor,
          object : BiometricPrompt.AuthenticationCallback() {
            override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
              super.onAuthenticationSucceeded(result)
              promise.resolve(
                AuthenticateResult(
                  success = true,
                  biometryType = primaryType,
                  error = null,
                  warning = null
                )
              )
            }

            override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
              super.onAuthenticationError(errorCode, errString)
              val mappedError = when (errorCode) {
                BiometricPrompt.ERROR_USER_CANCELED,
                BiometricPrompt.ERROR_NEGATIVE_BUTTON -> "userCanceled"
                BiometricPrompt.ERROR_CANCELED -> "systemCanceled"
                BiometricPrompt.ERROR_LOCKOUT -> "lockout"
                BiometricPrompt.ERROR_LOCKOUT_PERMANENT -> "lockoutPermanent"
                BiometricPrompt.ERROR_NO_BIOMETRICS -> "notEnrolled"
                BiometricPrompt.ERROR_HW_NOT_PRESENT,
                BiometricPrompt.ERROR_HW_UNAVAILABLE -> "hardwareUnavailable"
                BiometricPrompt.ERROR_NO_DEVICE_CREDENTIAL -> "passcodeNotSet"
                BiometricPrompt.ERROR_TIMEOUT -> "timeout"
                else -> "unknown"
              }

              promise.resolve(
                AuthenticateResult(
                  success = false,
                  biometryType = primaryType,
                  error = mappedError,
                  warning = errString.toString()
                )
              )
            }

            override fun onAuthenticationFailed() {
              super.onAuthenticationFailed()
              // BiometricPrompt keeps session active for retries until user cancels or lockout
            }
          }
        )

        biometricPrompt.authenticate(promptInfo)
      } catch (e: Throwable) {
        promise.resolve(
          AuthenticateResult(
            success = false,
            biometryType = primaryType,
            error = "unknown",
            warning = e.message
          )
        )
      }
    }

    return promise
  }
}

File: ios/NitroBiometrics.swift
import Foundation
import LocalAuthentication
import NitroModules

class NitroBiometrics: HybridNitroBiometricsSpec {
  
  public func canAuthenticate(allowDeviceCredentials: Bool?) throws -> Promise<BiometricsStatus> {
    let promise = Promise<BiometricsStatus>()
    
    let context = LAContext()
    let policy: LAPolicy = (allowDeviceCredentials == true) ? .deviceOwnerAuthentication : .deviceOwnerAuthenticationWithBiometrics
    
    var laError: NSError?
    let canEvaluate = context.canEvaluatePolicy(policy, error: &laError)
    
    let biometryType: BiometryType
    var biometryTypes: [BiometryType] = []
    
    switch context.biometryType {
    case .faceID:
      biometryType = .faceId
      biometryTypes = [.faceId]
    case .touchID:
      biometryType = .touchId
      biometryTypes = [.touchId]
    case .opticID:
      biometryType = .opticId
      biometryTypes = [.opticId]
    case .none:
      biometryType = .none
    @unknown default:
      biometryType = .none
    }
    
    var isDeviceSecure = false
    var passcodeError: NSError?
    if context.canEvaluatePolicy(.deviceOwnerAuthentication, error: &passcodeError) {
      isDeviceSecure = true
    } else if let pErr = passcodeError, pErr.code != LAError.passcodeNotSet.rawValue {
      isDeviceSecure = true
    }
    
    let enrolled = (context.biometryType != .none) && (laError?.code != LAError.biometryNotEnrolled.rawValue)
    
    var mappedError: String? = nil
    if let error = laError {
      mappedError = mapLAErrorCode(error.code)
    }
    
    let status = BiometricsStatus(
      isAvailable: canEvaluate,
      biometryType: biometryType,
      biometryTypes: biometryTypes,
      enrolled: enrolled,
      isDeviceSecure: isDeviceSecure,
      error: mappedError
    )
    
    promise.resolve(withResult: status)
    return promise
  }
  
  public func authenticate(options: AuthenticateOptions) throws -> Promise<AuthenticateResult> {
    let promise = Promise<AuthenticateResult>()
    let context = LAContext()
    
    if let fallbackTitle = options.fallbackButtonText {
      context.localizedFallbackTitle = fallbackTitle
    }
    if let cancelTitle = options.cancelButtonText {
      context.localizedCancelTitle = cancelTitle
    }
    
    let policy: LAPolicy = (options.allowDeviceCredentials == true) ? .deviceOwnerAuthentication : .deviceOwnerAuthenticationWithBiometrics
    
    let biometryType: BiometryType
    switch context.biometryType {
    case .faceID:
      biometryType = .faceId
    case .touchID:
      biometryType = .touchId
    case .opticID:
      biometryType = .opticId
    case .none:
      biometryType = .none
    @unknown default:
      biometryType = .none
    }
    
    let prompt = options.promptMessage
    
    context.evaluatePolicy(policy, localizedReason: prompt) { [weak self] success, error in
      guard let self = self else { return }
      
      if success {
        let result = AuthenticateResult(
          success: true,
          biometryType: (biometryType != .none) ? biometryType : nil,
          error: nil,
          warning: nil
        )
        promise.resolve(withResult: result)
      } else {
        var errorString = "unknown"
        if let laError = error as? LAError {
          errorString = self.mapLAErrorCode(laError.code.rawValue)
        } else if let nsError = error as? NSError {
          errorString = self.mapLAErrorCode(nsError.code)
        }
        
        let result = AuthenticateResult(
          success: false,
          biometryType: (biometryType != .none) ? biometryType : nil,
          error: errorString,
          warning: nil
        )
        promise.resolve(withResult: result)
      }
    }
    
    return promise
  }
  
  private func mapLAErrorCode(_ code: Int) -> String {
    switch code {
    case LAError.authenticationFailed.rawValue:
      return "failed"
    case LAError.userCancel.rawValue:
      return "userCanceled"
    case LAError.userFallback.rawValue:
      return "userFallback"
    case LAError.systemCancel.rawValue:
      return "systemCanceled"
    case LAError.passcodeNotSet.rawValue:
      return "passcodeNotSet"
    case LAError.biometryNotAvailable.rawValue:
      return "hardwareUnavailable"
    case LAError.biometryNotEnrolled.rawValue:
      return "notEnrolled"
    case LAError.biometryLockout.rawValue:
      return "lockout"
    case LAError.appCancel.rawValue:
      return "appCanceled"
    case LAError.invalidContext.rawValue:
      return "invalidContext"
    case LAError.notInteractive.rawValue:
      return "notInteractive"
    default:
      return "unknown"
    }
  }
}
